### ISSUE TITLE
Best way to avoid some warnings?

### ISSUE THREAD

**Issue by DeybisMelendez** [godotengine/godot#27576]
**Godot version:**
Godot 3.1 Estable

**OS/device including version:**
Windows 10.

**Issue description:**
I would like to know what the objective is and what some of the warnings that come in godot 3.1 help, I have asked in some places before and most recommend "deactivate the warnings" so I ask.

I am not a professional programmer and I am always open to learn new things, so I would like you to recommend how I can "fix" the warnings.

I leave some examples in screenshot.

I apologize for my bad English, I speak Spanish.

---------

Me gustaría saber cual es el objetivo y en qué ayudan algunas de las advertencias que vienen en godot 3.1, he preguntado en algunos sitios antes y la mayoría recomienda "desactivar las advertencias" por eso pregunto.

No soy un programador profesional y yo estoy siempre abierto a aprender cosas nuevas, por lo que quisiera que me recomendaran cómo puedo "arreglar" las advertencias.

Dejo algunos ejemplos en screenshot.

Pido disculpas por mi mal inglés, yo hablo español.

**Steps to reproduce:**
you can see the warnings of which I have doubts in the images.

--------

Puedes ver las advertencias de las que tengo dudas en las imágenes

**Minimal reproduction project:**
I do not think it's necessary
![Screenshot (87)](https://user-images.githubusercontent.com/40205683/55343142-ba91e900-5467-11e9-94db-c25b61ebc759.png)

**Comment by girng** [comment]
![image](https://user-images.githubusercontent.com/45474847/55344073-47857400-5461-11e9-9463-e9cf89cda4fe.png)

**Comment by DeybisMelendez** [comment]
> 
> 
> ![image](https://user-images.githubusercontent.com/45474847/55344073-47857400-5461-11e9-9463-e9cf89cda4fe.png)
> 
> :)

Hello, I want to fix the warnings not ignore them, I want to know why they are there, if everyone ignores them then they should think about eliminating them in the future.

**Comment by cbscribe** [comment]
Warnings are information, highlighting things that *might* be issues in your code. This information may or may not be important to you. If the warning "return value unused" isn't a problem for you, then you can safely ignore it. But for someone else, this might point out a problem they need to address. They are pretty clear about what they are addressing.

For example, if you're using a KinematicBody, you might have

    move_and_slide(velocity)

Which is going to produce a warning; "move_and_slide returns a value, but the value is never used." This highlights the fact that it should be 

    velocity = move_and_slide(velocity)

Because you don't want to throw away the returned velocity adjustment from the function.

To sum up: Ignore or "fix" warnings as needed.

**Comment by DeybisMelendez** [comment]
> 
> 
> Warnings are information, highlighting things that _might_ be issues in your code. This information may or may not be important to you. If the warning "return value unused" isn't a problem for you, then you can safely ignore it. But for someone else, this might point out a problem they need to address. They are pretty clear about what they are addressing.
> 
> For example, if you're using a KinematicBody, you might have
> 
> ```
> move_and_slide(velocity)
> ```
> 
> Which is going to produce a warning; "move_and_slide returns a value, but the value is never used." This highlights the fact that it should be
> 
> ```
> velocity = move_and_slide(velocity)
> ```
> 
> Because you don't want to throw away the returned velocity adjustment from the function.
> 
> To sum up: Ignore or "fix" warnings as needed.

If I understand. But for example, I have a project that started with Godot 3.0.6, and now it sends me between 30 and 40 warnings of this type, I think that's the reason why many people deactivate it, however, I feel that they bother more than what they help, at least, should add exceptions in some rules like the delta for example.

I understand the move_and_slide but not the example with the connect that I wrote, and therefore I ask the best way to "fix" that warning.

**Comment by ShyRed** [comment]
Check out the [documentation](https://docs.godotengine.org/en/3.1/classes/class_object.html#class-object-method-connect) on the `connect` method: its return value is an [error code](https://docs.godotengine.org/en/3.1/classes/class_@globalscope.html#enum-globalscope-error) that you can use to find out if your connect was successful.

In order to "fix" the warning you could compare the result of your `connect` call against the `OK` enum value, for example...

**Comment by cbscribe** [comment]
It is OK to disable warnings, not every one has to be "fixed". Note that you can click the yellow triangle at the bottom of the editor and ignore individual warning lines.

There has been discussion about muting warnings for built-in methods like `_process()`, so that you don't get that "'delta' isn't used" warning.

**Comment by DeybisMelendez** [comment]
> 
> 
> Check out the [documentation](https://docs.godotengine.org/en/3.1/classes/class_object.html#class-object-method-connect) on the `connect` method: its return value is an [error code](https://docs.godotengine.org/en/3.1/classes/class_@globalscope.html#enum-globalscope-error) that you can use to find out if your connect was successful.
> 
> In order to "fix" the warning you could compare the result of your `connect` call against the `OK` enum value, for example...

```
func _ready():
	if connect("body_entered", self, "on_enter") != OK:
		print(name + " no se conectó" )
```
Thank you very much! :D

**Comment by girng** [comment]
in the op, a "return variable" to check if a signal was **successfully connected** is about the same as doing:
```
var hello = "a string!"
if hello:
print("the variable hello was created, it's a string!")
```

when using `connect`, it's already implied the connection is going to be successful. if it can't connect to something, it will already display an error. wrapping if statements around connects work, but it's redundant, and will  create verbose  code that ultimately hinders  gdscript's simplistic nature.

@DeybisMelendez it's best to disable warnings and just code your game. your code will be fine, believe me 🥇

**Comment by DeybisMelendez** [comment]
> 
> 
> in the op, a "return variable" to check if a signal was **successfully connected** is about the same as doing:
> 
> ```
> var hello = "a string!"
> if hello:
> print("the variable hello was created, it's a string!")
> ```
> 
> when using `connect`, it's already implied the connection is going to be successful. if it can't connect to something, it will already display an error. wrapping if statements around connects work, but it's redundant, and will create verbose code that ultimately hinders gdscript's simplistic nature.
> 
> @DeybisMelendez it's best to disable warnings and just code your game. your code will be fine, believe me 🥇. a developer shouldn't even have to worry about those 2 warnings.

**Comment by akien-mga** [comment]
> it's best to disable warnings and just code your game. your code will be fine, believe me 🥇. a developer shouldn't even have to worry about those 2 warnings.

That's very bad advice, and unwelcome on the official issue tracker. It's fine to advise this on other community platforms if you think warnings are useful, but on this tracker we discuss how to make the engine better, not how to work around its flaws.

Warnings are useful and are not meant to be ignored. Some very specific warning occurrences, like the return value of `connevt` being discarded, need to be reviewed independently and we should see how to improve their usefulness, not hide them under the carpet.

**Comment by girng** [comment]
It's not bad advice at all, lol. There is no reason a developer needs to be told they are not using a delta. Or, that their `connect` method didn't connect. That's already displayed as an error.

Just because it's advice **you** don't like, doesn't mean it's bad.  But thanks for talking down to me for trying to help the OP, it's appreciated.

**Comment by akien-mga** [comment]
It's not about whether I like your advice or not. GDScript warnings are a feature of the engine. If some warnings are annoying for all users and always need to be disabled, that's a bug in said feature, and it needs to be fixed.

So it is useful that users do mention the warnings such as the `connect` one which does not seem to add much value to their scripts. It is true that `connect` returns a value, which is either `OK` or `ERR_INVALID_PARAMETERS`. When it's the latter, an error will indeed be printed in all cases. So yeah, manual handling of the return value is redundant here. But disabling warnings outright to silence this very specific case is not the right approach.

Possible fixes:
- Make `connect` return `void` (breaks compat)
- Add a whitelist of builtin methods that should not trigger this warning
- Drop this warning altogether (seems overkill, it's a useful warning as exemplified above with `move_and_slide`, or as can be seen in bigger projects when you forget to use the return value of your own-defined method).

**Comment by eon-s** [comment]
Warnings are kind of broken right now because the engine throws them for elements that are not used locally but by other scripts, and there is no way to avoid them safely like with underscores on local functions.

This kind of warnings can also cause a flood and useful message drop in extreme cases.

Even if disable warning is a bad advice, it is the most practical solution offered by Godot 3.1, something may need to be done to prevent users to decide to disable it (maybe a less annoying or more smart warning system).

Related #26056

**Comment by girng** [comment]
@akien-mga  I'm glad you agree with the redundancy! I was merely just giving my advice/opinion to the OP as  the other   users.  I didn't want to list all the things wrong, because this topic seemed like the OP just had  some **specific** questions about **their** warnings. 

 If it was an issue about how the warning system could be improved in general, I would have posted differently.

 I respect you, and your dedication to the project so much, that you saying **my advice is bad and unwelcome** hurt. It hurt hard. Honestly, I think it was completely uncalled for. I've been here nearly as long as you. Obviously you are more knowledgeable on the C++ and managerial side, and that is commendable. I'm just trying to make the experience to the godot developer  easier.

 I would never give advice out that could hurt a godot developer. This warning system can be a pain the butt sometimes. A godot developer should **never** be worried about an "unused delta in _process/_physics_process".  That kind of warning is  so insignificant to the developer, it causes more harm than good!  I have the utmost respect to vnen, you and the other contributors. Please don't ever think I do not. We are into this together.

**Comment by CowThing** [comment]
This is an issue I've seen a lot of people talk about, annoying warnings that aren't really a big deal in most cases. I think I saw this suggestion in another issue, but an idea to fix this problem could be to have multiple preset warning levels that the user can choose from (in addition to selecting their own custom list of warnings like is possible now). The warning levels would just change enabled warnings to some presets. By default it should start on a lower warning level that has the more annoying warnings disabled. The way I see it, warnings are useful but all of them don't need to be enabled all the time. Some of the warnings could be enabled only as a last resort when you can't figure out why a piece of code isn't working, and then they might highlight what is causing the problem.

Personally I disable the following warnings:

- **Unused Class Variable** - Plenty of times I have a class variable that isn't actually used within the class itself, but is read by other objects in the game. This is especially a problem for custom Resource scripts, because they are used to store data, so most of their class variables aren't used at all within the script.
-  **Unused Argument** - As mentioned above, when using `_process(delta)` without needing delta. This warning can be fixed already by changing `delta` to `_delta`, because a leading underscore silences it. But it's just more work adding underscores to every unused variable so I'd rather keep this warning disabled.
- **Unused Signal** - In a few cases I have a class that has a signal, but that signal is emitted from somewhere else instead of within the class itself. Another case is sometimes I'll use `call_deferred("emit_signal", "my_signal")`, which doesn't register as the signal being emitted, even though it is.
- **Return Value Discarded** - As mentioned above, there's plenty of times where I don't actually care if a return value is discarded because the function is doing something, and is only returning an error or a bool to notify if it succeeded.

Those four warnings are the only ones that have ever really bothered me. But every other warning that is enabled by default is actually useful and has helped me catch problems. This is why I think it's bad for people to outright disable all warnings, because many of them are useful and aren't annoying.

**Comment by Calinou** [comment]
> **Unused Argument** - As mentioned above, when using _process(delta) without needing delta. This warning can be fixed already by changing delta to _delta, because a leading underscore silences it. But it's just more work adding underscores to every unused variable so I'd rather keep this warning disabled.

On the other hand, the unused argument warning has helped me catch bugs and save hours of debugging, as it can be very difficult to see whether you're actually using the right variable in a method.

Consider something like this (this is a contrived example, but I've witnessed it in real-world projects):

```gdscript
var some_var = 2

# The function below uses a similarly-named (but not identical) parameter name as the member variable above
func some_func(some_var2 = 4):
    return some_var

func _ready():
    # "Why do I get 2 here?"
    print(some_func())
```

**Comment by akien-mga** [comment]
@girng I'm sorry that my comment hurt your feelings, this was not my intention.

I mainly reacted because OP started an issue with this premise:

> I would like to know what the objective is and what some of the warnings that come in godot 3.1 help, I have asked in some places before and most recommend "deactivate the warnings" so I ask.

Which shows that they know that warnings can be deactivated, but they are willing to go further in the discussion. From there, it makes sense for us to see why these warnings are raised, why they are problematic, and how to solve that *for everyone*, not on a per-user basis by disabling warnings which are meant to be useful.

**Comment by LinuxUserGD** [comment]
I think that warnings about arguments of functions in the engine should not appear as they can't be fixed, e. g. func _process(_delta) instead of func _process(delta).

**Comment by Calinou** [comment]
@LinuxUserGD `func _process(_delta)` is the preferred way to ignore the unused argument, it should work just fine.

**Comment by DeybisMelendez** [comment]
> 
> 
> @LinuxUserGD `func _process(_delta)` is the preferred way to ignore the unused argument (as it doesn't require writing a comment to ignore the warning), it should work just fine.

As for the `_process(_delta)` I think it should not be the solution compared to other engines that can ignore their deltas, like the case of Love2D where you can write simply `function love.update()` and not `function love.update(dt)`.

I think it really is a godot error that the warning appears in the `delta` and we are forced to write `_delta`. At least, we should be able to call the function without that. `func _process()`

----------------------

En cuanto al `_process(_delta)` me parece que no debería ser la solución en comparación con otros engines que si se pueden ignorar sus deltas, como el caso de Love2D donde puedes escribir simplemente `function love.update()` y no `function love.update(dt)`.

Creo que realmente es un error de godot que la advertencia aparezca en el delta y nosotros estemos obligados a escribir _delta. Al menos, deberíamos poder llamar la función sin eso. `func _process()`.

**Comment by golddotasksquestions** [comment]
Nobody seems to discussing the fact [that these warnings HIDE errors](https://github.com/godotengine/godot/issues/21896#issuecomment-480632953), making debugging impossible while warnings are enabled.

**Comment by amisner2k** [comment]
Hi, I wanted to drop a suggestion for the OP if they or others are still interested.

At least for the unused return value warning, the `assert()` can serve as a nice way to "consume" the value which has the added benefit of being stripped in release builds as well as providing an actual way to use the return value in a meaningful way. Example:

`assert(connect("changed", self, "update") == OK)`

The code still reads well, is concise, isn't much more to type, and actually conveys an intention that you want the connection result to be OK.

Anyway, hope this helps.

**Comment by Calinou** [comment]
Code within an `assert()` statement should never include side effects. Otherwise, these will never be performed in a release build, which will break your game :slightly_smiling_face:

**Comment by amisner2k** [comment]
but......in the case of connect...if you actually don't care about the result (like if it's OK, then everything's fine, then it should still work fine regardless right?

**Comment by amisner2k** [comment]
oh......in release builds that entire line would be stripped out wouldn't it.....dangit!

**Comment by amisner2k** [comment]
I was looking through the @GDScript globals in the docs and couldn't find any other `void` returning functions one could use for my idea. Would it be feasible to add a new global `void` method to GDScript to basically allow programmers to "swallow" return values for functions that they don't care about to make that warning disappear? Something like a `use(...)` so you could have:

`use(connect("changed", self, "update"))`

Just a thought.

**Comment by CowThing** [comment]
I don't think there needs to be a function that has no use beyond stopping a warning. It would make code harder to reason with for no reason other than not wanting to disable the warning. Calling a function and just not using the return value is easier to understand.

I still think having preset warning levels would help users that don't know what the warnings are. That way they can set it to low, medium, or high warning levels. Having in-engine descriptions that explain what each warning is for would help as well. Also maybe there should be a button to globally ignore a warning which will automatically disable that warning type in the project settings, here's a mockup of what that could look like.

![warnings](https://user-images.githubusercontent.com/11450875/65165041-fae02600-da0b-11e9-92af-2116121c21f8.png)

**Comment by amisner2k** [comment]
Agreed. Well the real confusion comes from when you are just *calling* the function without assigning its return to a variable.

IMHO:
`connect(...)` should not throw a warning
but
`var error = connect(...)` *should* throw a warning if I don't use `error` elsewhere in code.

So perhaps, GDScript parser (or whatever backend system handles this) could be improved to not report this particular "unused return value" in the case where the value is not actually assigned to a variable.

**Comment by Calinou** [comment]
@amisner2k This can still be the source of bugs, so I'd prefer to keep it the current way.

**Comment by eon-s** [comment]
The main problem is that warnings and errors share space and, as mentioned before, harmless warnings hide errors unless you disable them, making this system a big problem in that sense.

Maybe warnings and erros could be shown in different tabs?

**Comment by amisner2k** [comment]
Fair point @Calinou 
I think as @eon-s mentions, being able to somehow separate the warnings into their own view/list so that they don't overflow and spam the Error List would be the best way to "avoid some warnings" as the title of this Issue suggests. Having a bunch of warnings you don't care about seeing will be much less of an issue if they were kept separated from actual errors (especially since the log can become "full")

I think the feature that @CowThing describes would be a great feature to add (and should get its own Proposal and Issue created).

**Comment by Calinou** [comment]
@eon-s We should have a way to filter out specific message categories in the Errors tab, like in most Web browsers' developer tools:

![image](https://user-images.githubusercontent.com/180032/65177718-8df27d80-da57-11e9-8de9-79ac506c3a3d.png)

**Comment by amisner2k** [comment]
@Calinou , so back to the `connect()` thing.  What would be your recommended approach to avoid "The function 'connect()' returns a value, but this value is never used." warning?

What would your idea use of `connect()` be?

**Comment by mosmetro** [comment]
> What would your ideal use of `connect()` be to avoid a warning?

Something like this:
var _ = connect()
Underscore would indicates ignored value or unused variable. Just like in Swift & Lua.

**Comment by Calinou** [comment]
@mosmetro This is already possible, but you need to use at least two underscores, as GDScript doesn't allow single underscore identifiers. (This is probably a bug.)

**Comment by norton-corbett** [comment]
The language Nim has a `discard` statement i.e.

`discard connect(...)`

Which is nice because it's more explicit in purpose than `var __ = connect(...)` or whatever

**Comment by pycbouh** [comment]
It was mentioned that hiding warnings is not a valid solution to these types of problems, and yet there is an option to do so in engine. So there are people who are going to use it, even if it is discouraged, and there are cases for disabling warnings. Like, in my case, I have disabled a bunch of them: some like with `connect`, because it is generally not useful to get such warnings, others are disabled because I do not need to have them "on" constantly while developing, as things tend to be left unfinished for some time.

But benefits of having warnings should not be wasted even in case of having them generally disabled. How about a way to run static analysis on demand with a desired warnings and errors configured? Like, you disabled some inspection "live", but have a button or a build step to check all files for issues?

**Comment by RabdPnguin** [comment]
> > What would your ideal use of `connect()` be to avoid a warning?
> 
> Something like this:
> var _ = connect()
> Underscore would indicates ignored value or unused variable. Just like in Swift & Lua.

That's how C# does it too. It's called a discard. And since Godot supports C# may as well use a similar construct.

The only thing I don't like about it is underscore `_` is its valid variable name, so it could be confusing to the reader whether it's a variable or a discard.

With the above in mind (and the way C# does it) is you wouldn't have the keyword var.
`_ = connect()`

`var` would indicate you're assigning the value to the variable `_`, drop the var and it's used as a discard.  It could work for args too `do_something(_, prop_i_want, _)` or `_process(_)` Better yet, allow function overloading =p

**Comment by eon-s** [comment]
This issue is already solved in the documentation of the [warning system](https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/warning_system.html)
(the underscore thing is just mentioned somewhere else I think)

**Comment by Calinou** [comment]
Closing per @eon-s' comment.
